modules/evm/src/lib.rs

```diff
@@ -1025,42 +875,27 @@ impl<T: Config> Module<T> {
 		)
 	}
 
-	fn do_update_maintainer_storage_usage(
-		maintainer: &EvmAddress,
+	fn do_update_storage_usage(
+		origin: &EvmAddress,
 		pre_storage_usage: u32,
 		current_storage_usage: u32,
+		storage_limit: u32,
 	) -> DispatchResult {
-		// get maintainer quota and pay for the additional_storage
-		<Accounts<T>>::mutate(
-			maintainer,
-			|maybe_maintainer_account_info| -> Result<(), DispatchError> {
-				if let Some(AccountInfo {
-					storage_quota: maintainer_storage_quota,
-					storage_usage: maintainer_storage_usage,
-					..
-				}) = maybe_maintainer_account_info.as_mut()
-				{
-					if let Some(delta) = current_storage_usage.checked_sub(pre_storage_usage) {
-						*maintainer_storage_usage = maintainer_storage_usage
-							.checked_add(delta)
-							.ok_or(Error::<T>::NumOutOfBound)?;
-					} else if let Some(delta) = pre_storage_usage.checked_sub(current_storage_usage) {
-						*maintainer_storage_usage = maintainer_storage_usage
-							.checked_sub(delta)
-							.ok_or(Error::<T>::NumOutOfBound)?;
-					}
+		if let Some(delta) = current_storage_usage.checked_sub(pre_storage_usage) {
+			if delta > storage_limit {
+				return Err(Error::<T>::StorageExceedsStorageLimit.into());
+			}
 
-					if *maintainer_storage_usage > *maintainer_storage_quota {
-						return Err(Error::<T>::StorageQuotaNotEnough.into());
-					}
+			let reserve_amount = T::StorageDepositPerByte::get().saturating_mul(delta.into());
+			let origin_account = T::AddressMapping::get_account_id(&origin);
+			T::Currency::reserve(&origin_account, reserve_amount)?;
```

Comment:
the reserve & unreserve may happen on different accounts. e.g. one account store 100 bytes and another account remove those bytes

so need to transfer the deposit from origin to contract, and reserve on the contract account.

on refund, use repatriate_reserved to unreserve and send refund to origin. ignore the error if no enough fund to unreserve